utils/mof_checker/__init__.py

```diff
@@ -0,0 +1,123 @@
+    "no_overcoordinated_carbon",
+    "no_overcoordinated_hydrogen",
+    "no_overcoordinated_nitrogen",
+    "no_undercoordinated_nitrogen",
+    "no_undercoordinated_rare_earth",
+    "no_floating_molecule",
+    "no_false_terminal_oxo",
+]
+
+class MofCheckerWidget(ipw.VBox):
+    """Widget that allows to check MOF structure correctness."""
+    structure = Instance(Atoms, allow_none=True)
+    selection = List(Int)
+
+    def __init__(self, title=''):
+        self.title = title
+        button_check = ipw.Button(description="Analyse structure")
+        button_check.on_click(self.check_structure)
+        self.checks = []
+        self._output = ipw.Output()
+        self.missing_h_button = ipw.Button(description="Add", layout={"width": "initial"})
+
+        super().__init__(children=[button_check,
+                                   self._output,
+                                  ])
+    
+    def check_structure(self, _=None):
+        
+        if self.structure is None:
+            return
+
+        self.mfchk = MOFChecker.from_ase(self.structure, primitive=False)
+        self.checks = []
+        if self.structure:
+            self.checks = [ check for (check_name, check) in self.mfchk.checks.items() if check_name in ENABLED_CHECKS]
+        
+        issue_found = False
+        
+        with self._output:
+            clear_output()
+            for check in self.checks:
+                if not check.is_ok:
+                    issue_found = True
+                    try:
+                        selection = list(chain.from_iterable(check.flagged_indices))
+                    except TypeError:
+                        selection = check.flagged_indices
+                    button = ipw.Button(description="Select", layout={"width": "initial"})
+                    button.on_click(fct.partial(self._select_atoms, selection=selection))
+                    #text = ipw.HTML("Found " + check.name.lower() + ": " + ', '.join(map(str, selection)))
+                    text = ipw.HTML("Found " + check.name.lower())
+                    display(ipw.HBox([text, button]))
+                    
+            if self.mfchk.undercoordinated_c_candidate_positions + self.mfchk.undercoordinated_n_candidate_positions:
+                self.missing_h_button.disabled = False
+                self.missing_h_button.on_click(self.add_missing_hydrogens)
+                display(ipw.HBox([ipw.HTML("Missing hydrogens found"), self.missing_h_button]))
+            
+            if not issue_found:
+                display(ipw.HTML("No issues found \u2705"))
+            
+    def _select_atoms(self, _=None, selection=None):
+        self.selection = selection
+
+    def add_missing_hydrogens(self, _=None):
+        """Add atoms."""
+        atoms = self.structure.copy()
+        selection = self.selection
+
+        h_positions = self.mfchk.undercoordinated_c_candidate_positions + self.mfchk.undercoordinated_n_candidate_positions
```

Comment:
I'll double check if the list is always flat. I think this is atm not good in `mofchecker` as it might return you a list of lists of positions in case you have multiple kinds (sp2, sp3...) missing H

Reply:
>  it might return you a list of lists of positions in case you have multiple kinds (sp2, sp3...) missing H

eh, maybe this level of detail isn't really needed.